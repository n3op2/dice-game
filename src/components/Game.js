import React from 'react';

import Header from './Header';
import Player from './Player';
import { Button, Container, AppContext, Table } from './shared';

const renderGameElements = (props) => {
    const { history, player, computer, started, update } = props;

    if (!started) return <Button
        onClick={() => update({ started: true })}
        name="start"
    />

    return (
        <Container style={{ display: 'flex', flexDirection: 'row', height: 300 }}>
            <Table history={history} />
            <Player {...player} />
            <Player {...computer} />
            <Container>{started && <Button
                onClick={() => update({
                    playing: true,
                    status: 'rolling dices...',
                    count: props.count + 1,
                })}
                name="roll dice"
            />}</Container>
        </Container>
    );
};

const updateScore = (players, update, log) => {
    log.add(players);
    if (new Set(players.map((p) => p.val)).size === 1)
        return update({ status: "It's a draw!", playing: false });

    const winner = players.reduce((leading, next) =>
        leading.val > next.val ? leading : next
    );
    winner.score++;

    return update({
        status: `${winner.name} has won this game!`,
        [winner.name]: winner,
        playing: false,
    });
};

const Game = () => {
    const state = React.useContext(AppContext);
    const { player, computer, update, log } = state;

    React.useEffect(() => {
        if (!player.rolling && !computer.rolling) {
            updateScore([player, computer], update, log);
        }
    }, [player, computer]);

    return (
        <Container>
            <Header />
            {renderGameElements(state)}
            <div>footer: {state.count}</div>
        </Container>
    );
};

export default Game;
