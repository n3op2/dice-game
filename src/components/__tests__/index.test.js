import React from 'react';
import { shallow } from 'enzyme';
import Game from '../Game';

const mockContext = {
    playing: false,
    started: true,
    history: [],
    count: 0,
    status: 'not played',
    player: { score: 0, val: 0, name: 'player', rolling: false },
    computer: { score: 0, val: 0, name: 'computer', rolling: false },
}

describe('Game Component', () => {
    let wrapper;

    beforeEach(() => {
        React.useContext = jest.fn().mockReturnValue(mockContext);
        wrapper = shallow(<Game />);
    });

    test('renders a component', () => {
        expect(wrapper).toMatchSnapshot();
    });

});
