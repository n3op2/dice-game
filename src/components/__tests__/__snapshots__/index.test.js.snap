// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Component renders a component 1`] = `
<Container>
  <Header />
  <Container
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "height": 300,
      }
    }
  >
    <Table
      history={Array []}
    />
    <Player
      name="player"
      rolling={false}
      score={0}
      val={0}
    />
    <Player
      name="computer"
      rolling={false}
      score={0}
      val={0}
    />
    <Container>
      <Button
        name="roll dice"
        onClick={[Function]}
      />
    </Container>
  </Container>
  <div>
    footer: 
    0
  </div>
</Container>
`;
